sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -1083,6 +1124,69 @@ def test_uint8_predict(Est):
 
 
 # TODO(1.3): Remove
+@pytest.mark.parametrize(
+    "interaction_cst, n_features, result",
+    [
+        (None, 931, None),
+        ([{0, 1}], 2, [{0, 1}]),
+        ([(1, 0), [5, 1]], 6, [{0, 1}, {1, 5}, {2, 3, 4}]),
+    ],
+)
+def test_check_interaction_cst(interaction_cst, n_features, result):
+    """Check that _check_interaction_cst returns the expected list of sets"""
+    est = HistGradientBoostingRegressor()
+    est.set_params(interaction_cst=interaction_cst)
+    assert est._check_interaction_cst(n_features) == result
+
+
+def test_interaction_cst_numerically():
+    """Check that interaction constraints have no forbidden interactions."""
+    rng = np.random.RandomState(42)
+    n_samples = 1000
+    X = rng.uniform(size=(n_samples, 2))
+    # Construct y with a strong interaction term
+    # y = x0 + x1 + 5 * x0 * x1
+    y = np.c_[X, 5 * X[:, 0] * X[:, 1]].sum(axis=1)
+
+    est = HistGradientBoostingRegressor()
+    est.fit(X, y)
+    est_no_interactions = HistGradientBoostingRegressor(interaction_cst=[{0}, {1}])
+    est_no_interactions.fit(X, y)
+
+    delta = 0.25
+    # Make sure we do not extrapolate out of the training set as tree-based estimators
+    # are very bad in doing so.
+    X_test = X[(X[:, 0] < 1 - delta) & (X[:, 1] < 1 - delta)]
+    X_delta_0 = X_test + [delta, 0]
+    X_delta_1 = X_test + [0, delta]
+    X_delta_0_1 = X_test + [delta, delta]
+
+    # Note: For the y from above as a function of x0 and x1, we have
+    # y(x0+d, x1+d) = y(x0, x1) + 5 * d * (2/5 + x0 + x1) + 5 * d**2
+    # y(x0+d, x1)   = y(x0, x1) + 5 * d * (1/5 + x1)
+    # y(x0,   x1+d) = y(x0, x1) + 5 * d * (1/5 + x0)
+    # Without interaction constraints, we would expect a result of 5 * d**2 for the
+    # following expression, but zero with constraints in place.
+    assert_allclose(
+        est_no_interactions.predict(X_delta_0_1)
+        + est_no_interactions.predict(X_test)
+        - est_no_interactions.predict(X_delta_0)
+        - est_no_interactions.predict(X_delta_1),
+        0,
+        atol=1e-12,
+    )
+
+    # Correct result of the expressions is 5 * delta**2. But this is hard to achieve by
+    # a fitted tree-based model. The expression should, however, at least be positive!
```

Comment:
I'm wondering if we should include this test here. It's not really checking the interaction constraints.

Also it fails when there are not enough trees:

```python
def test_interaction_cst_numerically_v2():
    """Check that interaction constraints have no forbidden interactions."""
    rng = np.random.RandomState(42)
    n_samples = 1000
    X = rng.uniform(size=(n_samples, 2))
    y = np.hstack((X, 5 * X[:, [0]] * X[:, [1]])).sum(axis=1)

    est = HistGradientBoostingRegressor(max_iter=, random_state=0)
    est.fit(X, y)

    delta = 0.25
    X_test = X[(X[:, 0] < 1 - delta) & (X[:, 1] < 1 - delta)]
    X_delta_0 = X_test + [delta, 0]
    X_delta_1 = X_test + [0, delta]
    X_delta_0_1 = X_test + [delta, delta]
    assert np.all(
        est.predict(X_delta_0_1)
        + est.predict(X_test)
        - est.predict(X_delta_0)
        - est.predict(X_delta_1)
        > 0.01
    )
```

Reply:
That is a bit the point: Tree based models are very bad in capturing 2-dimensional (or higher) structures which are not coordinate/axis aligned (not well approximated with rectangles/hypercubes).

While I do not insist on this test, I find it very neat. It is cheap to compute and needs only these few lines of code if placed here. Semantically, it's not the best place.

Reply:
I see. Okay, let's keep it.

Reply:
From this conversation, what do you think of adding "with 100 iterations"?

```suggestion
    # Correct result of the expressions is 5 * delta**2. But this is hard to achieve by
    # a fitted tree-based model. However, with 100 iterations the expression should
    # at least be positive!
```